lib/xenops_server.ml

```diff
@@ -2514,9 +2514,28 @@ and perform_exn ?subtask ?result (op : operation) (t : Xenops_task.task_handle)
   in
   one op
 
+and verify_power_state op =
```

Comment:
Is the power state up-to-date here, or do we need to process any pending state updates first?
i.e. if we have queued operations like: $something, $something, $something, VM.reboot, $something, $something, check_state, $something. Then we should probably execute a check_state before VM.reboot anyway (even though it was queued later) because we can only reboot a VM based on its current power state.
Should we always queue an op that updates the power state based on the state of the Xen domain just prior to start/reboot?

Reply:
I like the idea to ask for the power state before reboot explicitly.

Reply:
That is precisely what this is doing: `verify_power_state` is called right before executing an operation after popping it from the queue. It uses `B.VM.get_state` below to get the _current_ state of the domain, with info straight from xen (this is not cached).

Reply:
Good that it is not cached. `get_state` is a bit heavy-weight (it fetches whole xenstore subtree), but doing it just for start/reboot/resume hopefully won't introduce too many performance issues.